### PR TITLE
Verify Examples

### PR DESCRIPTION
**Pull request by matthew-hennefarth**
Addresses #1820 as well as tests failing: #1819, #1725, #1726. Would also prevent future issues from being created due to examples failing.

`verify_examples.py` is a script that can run the examples (or a subset) and verify if it terminates normally. I did not add any filtering options at the moment, but this could be added if necessary.

I would recommend to use it like:

```sh
# in the pyscf directory 
./tools/verify_examples.py examples 
```
Also, there is an option to run the examples in parallel using the `-j` or `--jobs` flag. This works like `make`. So:
```sh
./tools/verify_examples.py examples -j 8 
```
parallelize to 8 threads. Make sure to set `OMP_NUM_THREADS` to 1 otherwise there might be issues with the PySCF runtime.

Finally, for the future, it is possible to only run a subset of tests by just passing that directory. As an example:

```sh
tools/verify_examples.py examples/gto
```

would only run the examples under `pyscf/examples/gto`. 

I have not set up the github actions to run this automatically as there are a *lot* of examples that fail. Some also produce matplotlib plots as outputs. 

## Example Output
Here is an example of the output if I run it on the `examples/dft`:

```sh
(pyscf) pyscf [example-ci] > tools/verify_examples.py examples/dft -j 7

running 20 examples
[  5%]: dft/14-collinear_dks.py ... FAILED
[ 10%]: dft/00-simple_dft.py ... ok
[ 15%]: dft/14-collinear_gks.py ... FAILED
[ 20%]: dft/02-gks.py ... ok
[ 25%]: dft/11-grid_scheme.py ... ok
[ 30%]: dft/13-rsh_dft.py ... ok
[ 35%]: dft/16-dft_d3.py ... FAILED
[ 40%]: dft/10-glycine.py ... ok
[ 45%]: dft/20-density_fitting.py ... ok
[ 50%]: dft/21-x2c.py ... ok
[ 55%]: dft/24-define_xc_functional.py ... FAILED
[ 60%]: dft/30-ao_value_on_grid.py ... ok
[ 65%]: dft/23-decorate_scf.py ... ok
[ 70%]: dft/22-newton.py ... ok
[ 75%]: dft/24-custom_xc_functional.py ... ok
[ 80%]: dft/31-xc_value_on_grid.py ... FAILED
[ 85%]: dft/15-nlc_functionals.py ... ok
[ 90%]: dft/32-xcfun_as_default.py ... ok
[ 95%]: dft/12-camb3lyp.py ... ok
[100%]: dft/32-broken_symmetry_dft.py ... ok

failures:

---- dft/14-collinear_dks.py stderr ----
Traceback (most recent call last):
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/numint2c.py", line 301, in mcfun_eval_xc_adapter
    import mcfun
ModuleNotFoundError: No module named 'mcfun'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mhennefarth/Developer/pyscf/examples/dft/14-collinear_dks.py", line 17, in <module>
    mf.kernel()
  File "<string>", line 2, in kernel
  File "/Users/mhennefarth/Developer/pyscf/pyscf/scf/dhf.py", line 665, in scf
    = kernel(self, self.conv_tol, self.conv_tol_grad,
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/scf/dhf.py", line 67, in kernel
    = hf.kernel(mf, 1e-2, 1e-1,
      ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/scf/hf.py", line 129, in kernel
    vhf = mf.get_veff(mol, dm)
          ^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/gks.py", line 76, in get_veff
    n, exc, vxc = ni.get_vxc(mol, ks.grids, ks.xc, dm,
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/r_numint.py", line 467, in r_vxc
    eval_xc = ni.mcfun_eval_xc_adapter(xc_code)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/numint2c.py", line 303, in mcfun_eval_xc_adapter
    raise ImportError('This feature requires mcfun library.\n'
ImportError: This feature requires mcfun library.
Try install mcfun with `pip install mcfun`

---- dft/14-collinear_gks.py stderr ----
Traceback (most recent call last):
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/numint2c.py", line 301, in mcfun_eval_xc_adapter
    import mcfun
ModuleNotFoundError: No module named 'mcfun'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mhennefarth/Developer/pyscf/examples/dft/14-collinear_gks.py", line 17, in <module>
    mf.kernel()
  File "<string>", line 2, in kernel
  File "/Users/mhennefarth/Developer/pyscf/pyscf/scf/hf.py", line 1917, in scf
    kernel(self, self.conv_tol, self.conv_tol_grad,
  File "/Users/mhennefarth/Developer/pyscf/pyscf/scf/hf.py", line 129, in kernel
    vhf = mf.get_veff(mol, dm)
          ^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/gks.py", line 76, in get_veff
    n, exc, vxc = ni.get_vxc(mol, ks.grids, ks.xc, dm,
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/numint2c.py", line 653, in nr_vxc
    n, exc, vmat = self._gks_mcol_vxc(mol, grids, xc_code, dms, relativity,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/numint2c.py", line 125, in _gks_mcol_vxc
    eval_xc = ni.mcfun_eval_xc_adapter(xc_code)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/numint2c.py", line 303, in mcfun_eval_xc_adapter
    raise ImportError('This feature requires mcfun library.\n'
ImportError: This feature requires mcfun library.
Try install mcfun with `pip install mcfun`

---- dft/16-dft_d3.py stderr ----
Traceback (most recent call last):
  File "/Users/mhennefarth/Developer/pyscf/examples/dft/16-dft_d3.py", line 48, in <module>
    mf.kernel()
  File "<string>", line 2, in kernel
  File "/Users/mhennefarth/Developer/pyscf/pyscf/scf/hf.py", line 1917, in scf
    kernel(self, self.conv_tol, self.conv_tol_grad,
  File "/Users/mhennefarth/Developer/pyscf/pyscf/scf/hf.py", line 130, in kernel
    e_tot = mf.energy_tot(dm, h1e, vhf)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/scf/hf.py", line 304, in energy_tot
    e_disp = mf.get_dispersion()
             ^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/scf/dispersion.py", line 146, in get_dispersion
    from pyscf.dispersion import dftd3, dftd4
ModuleNotFoundError: No module named 'pyscf.dispersion'

---- dft/24-define_xc_functional.py stderr ----
Traceback (most recent call last):
  File "/Users/mhennefarth/Developer/pyscf/examples/dft/24-define_xc_functional.py", line 63, in <module>
    mf.kernel()
  File "<string>", line 2, in kernel
  File "/Users/mhennefarth/Developer/pyscf/pyscf/scf/hf.py", line 1917, in scf
    kernel(self, self.conv_tol, self.conv_tol_grad,
  File "/Users/mhennefarth/Developer/pyscf/pyscf/scf/hf.py", line 129, in kernel
    vhf = mf.get_veff(mol, dm)
          ^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/rks.py", line 80, in get_veff
    n, exc, vxc = ni.nr_rks(mol, ks.grids, ks.xc, dm, max_memory=max_memory)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/numint.py", line 1158, in nr_rks
    for i, ao, mask, wv in block_loop(ao_deriv):
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/numint.py", line 1138, in block_loop
    exc, vxc = ni.eval_xc_eff(xc_code, rho, deriv=1, xctype=xctype)[:2]
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/numint.py", line 2700, in eval_xc_eff
    out = self.eval_xc1(xc_code, rho, spin, deriv, omega)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mhennefarth/Developer/pyscf/pyscf/dft/libxc.py", line 1802, in _eval_xc1
    out = numpy.vstack(out)[:outlen]
          ^^^^^^^^^^^^^^^^^
  File "<__array_function__ internals>", line 200, in vstack
  File "/opt/homebrew/lib/python3.11/site-packages/numpy/core/shape_base.py", line 296, in vstack
    return _nx.concatenate(arrs, 0, dtype=dtype, casting=casting)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<__array_function__ internals>", line 200, in concatenate
ValueError: all the input array dimensions except for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 30056 and the array at index 6 has size 1

---- dft/31-xc_value_on_grid.py stderr ----
Traceback (most recent call last):
  File "/Users/mhennefarth/Developer/pyscf/examples/dft/31-xc_value_on_grid.py", line 81, in <module>
    rhoL, mL = r_numint.eval_rho(mol, aoL_value, dmLL)
    ^^^^^^^^
ValueError: too many values to unpack (expected 2)


failures:
    dft/14-collinear_dks.py
    dft/14-collinear_gks.py
    dft/16-dft_d3.py
    dft/24-define_xc_functional.py
    dft/31-xc_value_on_grid.py

example results: FAILED. 15 passed; 5 failed; 0 filtered out; finished in 43.81s

error: examples failed
```

Note: I don't have `mcfun` installed on my local machine here, but it illustrates the utility of the script.

### CODE REVIEW

**Comment by sunqm** [comment]
It's nice to have these additional tests in CI jobs. At present, it can be a manual job IMO.

**Comment by matthew-hennefarth** [comment]
> At present, it can be a manual job IMO.

It will also take some time to fix/cleanup the examples so that it can pass the CI pipeline. I will convert the PR from a Draft so that the script is available.

File: tools/verify_examples.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
+#!/usr/bin/env python
+
```
Comment (sunqm) [inline tools/verify_examples.py:2]: Add a short docstring to describe how this script can be used, such as
```
# in the pyscf directory 
./tools/verify_examples.py examples 
./tools/verify_examples.py examples/gto -j 8
```

Additionally, consider to cite this PR in the docstring.

**Review by sunqm — APPROVED** [review]

File: tools/verify_examples.py (hunk 1 above)
Comment (matthew-hennefarth) [reply tools/verify_examples.py:53]: Added docstring giving example usage and reference to this PR.

**Pull request merged**